esp-hal/src/rmt.rs

```diff
@@ -1017,62 +1021,74 @@ impl<Raw> SingleShotTxTransaction<'_, Raw>
-                    // Not sure that this can happen? In any case, be sure that we don't lock up
-                    // here in case it can.
-                    return Err((Error::TransmissionError, self.channel));
-                }
-            }
+
+        let status = raw.get_tx_status();
+        if status == Some(Event::Threshold) {
             raw.reset_tx_threshold_set();
 
-            // re-fill TX RAM
-            let ptr = unsafe { raw.channel_ram_start().add(self.ram_index) };
-            let count = self.remaining_data.len().min(memsize / 2);
-            let (chunk, remaining) = self.remaining_data.split_at(count);
-            for (idx, entry) in chunk.iter().enumerate() {
-                unsafe {
-                    ptr.add(idx).write_volatile(*entry);
+            if !self.remaining_data.is_empty() {
+                // re-fill TX RAM
+                let memsize = raw.memsize().codes();
+                let ptr = unsafe { raw.channel_ram_start().add(self.ram_index) };
+                let count = self.remaining_data.len().min(memsize / 2);
+                let (chunk, remaining) = self.remaining_data.split_at(count);
+                for (idx, entry) in chunk.iter().enumerate() {
+                    unsafe {
+                        ptr.add(idx).write_volatile(*entry);
+                    }
                 }
+
+                // If count == memsize / 2 codes were written, update ram_index as
+                // - 0 -> memsize / 2
+                // - memsize / 2 -> 0
+                // Otherwise, for count < memsize / 2, the new position is invalid but the new
+                // slice is empty and we won't use ram_index again.
+                self.ram_index = memsize / 2 - self.ram_index;
+                self.remaining_data = remaining;
+                debug_assert!(
+                    self.ram_index == 0
+                        || self.ram_index == memsize / 2
+                        || self.remaining_data.is_empty()
+                );
             }
+        }
+
+        status
+    }
 
-            // If count == memsize / 2 codes were written, update ram_index as
-            // - 0 -> memsize / 2
-            // - memsize / 2 -> 0
-            // Otherwise, for count < memsize / 2, the new position is invalid but the new
-            // slice is empty and we won't use ram_index again.
-            self.ram_index = memsize / 2 - self.ram_index;
-            self.remaining_data = remaining;
-            debug_assert!(
-                self.ram_index == 0
-                    || self.ram_index == memsize / 2
-                    || self.remaining_data.is_empty()
-            );
+    /// Check transmission status and write new data to the hardware if
+    /// necessary.
+    ///
+    /// Returns whether transmission has ended (whether successfully or with an
+    /// error). In that case, a subsequent call to `wait()` returns immediately.
+    pub fn poll(&mut self) -> bool {
```

Comment:
Maybe this should return `Event`, and then we can add helper methods to determine whether it's ready for new data?

Something like `pub fn poll(&mut self) -> Result<Event, Error> {`, and the on `Event` we can have helper methods like:

```rust
pub fn is_done(&self) -> bool {
  match self {
    Event::Blah => true,
  }
}
```

What do you think?

Reply:
Or is what you're proposing simply something along the lines of

```rust
pub struct TransactionState {
    done: bool,
}

impl TransactionState {
    fn is_done(&self) -> bool { self.done }
}

fn poll(&mut self) -> TransactionState { ... }

// user code
let txn = channel.transmit(&data)?;
loop {
    if txn.poll().is_done() { break; }
}
let channel = txn.wait().unwrap();
```

to make the meaning of the bool more explicit at the call site?

Reply:
How does that work when the API itself doesn't seem to allow pushing data to an ongoing transaction? How can you not have a short period when the peripheral is idle?

Reply:
> It also places the burden of chunking the data on user code

Why? The driver can do the same things as it does now, while hiding all the polling/waiting/extra objects behind a single "do your work" call. So long as there is no actual double buffering necessary, the benefit of the single shot transaction is that it lets the CPU do its thing - although I guess the polling requirement actually ends up being a burden and the user just ends up writing async code without actually writing async code.

Reply:
> What should the path forward for this PR be?

Just to be clear, I'm not challenging anything you do here, I'm trying to learn here.

Reply:
Thanks for the pointer @Dominaezzz, that should work very nicely for RMT as well :+1: I've prototyped something locally, will submit this in a later PR.

How do we proceed here? Merge as-is, or remove/mark as `#[doc(hidden)]` the `.poll()` method until the blocking interface has been re-designed (I'd rather not do that here; it would blow the PR too much).